juju/juju
# LXD disk and architecture constraints

Supercedes #12375

The same commits are cherry-picked on to 2.9 instead of develop (3.0), so that we can release patch changes in the imminent 
release.

Addresses https://bugs.launchpad.net/juju/+bug/1905827

## QA steps

For the MB/MiB fix:
- Bootstrap LXD
- `juju add-machine --constraints "mem=4G cores=2"`
- Using the instance ID from juju status: `lxc config show <ID> | grep limits` should show:
```
  limits.cpu: "2"
  limits.memory: 4096MiB
```

Root disk:
- `juju add-machine --constraints "root-disk=4G root-disk-source=default"`
- `lxc config show <ID>` should include:
```
devices:
  root:
    path: /
    pool: default
    size: 4096MiB
    type: disk
```

Arch:
- Launch an arm64 instance on AWS with a keypair for SSH.
- Modify the associated security group to allow SSH from your machine and all TCP traffic to 8443.
- SSH to the instance:
  - `sudo snap refresh lxd --channel latest/stable`
  - `lxd init`. use sensible defaults, the public DNS name for the cluster address, and a trust password.
- Add a new LXD cloud/credentials for this new LXD cluster.
- `GOARCH=arm64 bootstrap ... --build-agent --bootstrap-constraints "arch=arm64"` will hang waiting for an address (can't access the internal bridge network), but via SSH `lxc list` will show that we successfully asked the server for an instance of a different arch to our client.
